pkg/tfgen/docs.go

```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
-			continue
+	// Find code fences in the presented doc
+	type tfBlockIndex struct {
+		start       int
+		end         int
+		headerStart int // a value of -1 indicates there's no header
+	}
+
+	var inCodeBlock bool
+	codeFence := "```"
+	h2 := "##"
+	h3 := "###"
+	var codeIndices []tfBlockIndex
+	var currentBlock tfBlockIndex
+	currentBlock.headerStart = -1
+
+	for i := 0; i < (len(docs) - len(codeFence)); i++ {
+
+		if inCodeBlock == true {
+			//Our TF code block is over right when we see the closing code fence.
+			if docs[i:i+len(codeFence)] == codeFence {
+				currentBlock.end = i
+				codeIndices = append(codeIndices, currentBlock)
+				// reset
+				currentBlock.start = 0
+				currentBlock.end = 0
+				inCodeBlock = false
+			}
+		} else {
+			// Keep track of header locations. These should never be inside code blocks.
+			if docs[i:i+len(h2)] == h2 || docs[i:i+len(h3)] == h3 {
+				// We need to make sure we don't reread an H3 as H2 on the next pass
+				if !(currentBlock.headerStart+1 == i) {
+					currentBlock.headerStart = i
+				}
+			}
+			// if we aren't in a code block, a set of code fences signals the beginning of a code block.
+			if docs[i:i+len(codeFence)] == codeFence {
+				inCodeBlock = true
+				currentBlock.start = i
+			}
 		}
+	}
 
-		isImportSection := false
-		header, wroteHeader := section[0], false
-		isFrontMatter, isExampleUsage := !strings.HasPrefix(header, "## "), header == "## Example Usage"
+	// Now we know where the headers and code fences are.
+	textStart := 0
+	stripSection := false
+	stripSectionHeader := 0
+	for _, tfBlock := range codeIndices {
 
-		if stripSubsectionsWithErrors && header == "## Import" {
-			isImportSection = true
-			isFrontMatter = false
-			wroteHeader = true
-		}
+		// if the section has a header we append the header after trying to convert the code.
+		hasHeader := tfBlock.headerStart >= 0 && textStart < tfBlock.headerStart
 
-		sectionStart, sectionEnd := "", ""
-		if isExampleUsage {
-			sectionStart, sectionEnd = "{{% examples %}}\n", "{{% /examples %}}"
+		// append non-code text to output
+		if !stripSection {
+			if hasHeader {
+				fprintf(output, docs[textStart:tfBlock.headerStart])
+			} else {
+				fprintf(output, docs[textStart:tfBlock.start])
+			}
+		} else {
+			// if we are stripping this section and still have the same header, we append nothing and skip to the next
+			// code block.
+			if stripSectionHeader == tfBlock.headerStart {
+				textStart = tfBlock.end + len(codeFence)
+				continue
+			}
+			if stripSectionHeader < tfBlock.headerStart {
+				stripSection = false
+			}
 		}
+		// find the actual start index of the code
+		nextNewLine := strings.Index(docs[tfBlock.start:tfBlock.end], "\n")
+		if nextNewLine == -1 {
+			// write the line as-is; this is an in-line fence
+			fprintf(output, docs[tfBlock.start:tfBlock.end]+"```")
+			writeTrailingNewline(output)
+		} else {
+			syntaxHighlight := docs[tfBlock.start : tfBlock.start+nextNewLine+1]
```

Comment:
Can we call this something more descriptive?
```suggestion
			codeFence := docs[tfBlock.start : tfBlock.start+nextNewLine+1]
```